marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -288,3 +331,145 @@ constantValue (Term (SubValue a b) _) = do
+                    || c
+                    == (codePointFromChar '\r')
+                    || c
+                    == (codePointFromChar ' ')
+                    || c
+                    == (codePointFromChar '\t')
+                )
+          )
+          token
+    in
+      [ { startColumn: column
+        , startLineNumber: row
+        , endColumn: column + (length word)
+        , endLineNumber: row
+        , message: message
+        , severity: markerSeverity "Error"
+        , code: ""
+        , source: ""
+        }
+      ]
+  Right state ->
+    let
+      holesMarkers = state ^. (_holes <<< to holesToMarkers)
+
+      warningsMarkers = state ^. (_warnings <<< to Set.toUnfoldable <<< to (map warningToMarker))
+    in
+      holesMarkers <> warningsMarkers
+
+-- other types of warning could do with being refactored to a Warning ADT first so we don't need to repeat ourselves
+holesToMarkers :: Holes -> Array IMarkerData
+holesToMarkers (Holes holes) =
+  let
+    (allHoles :: Array _) = Set.toUnfoldable $ fold $ Map.values holes
+  in
+    foldMap holeToMarkers allHoles
+
+holeToMarker :: MarloweHole -> Map String (Array Argument) -> String -> IMarkerData
+holeToMarker hole@(MarloweHole { name, marloweType, row, column }) m constructorName =
+  { startColumn: column
+  , startLineNumber: row
+  , endColumn: column + (length name) + 1
+  , endLineNumber: row
+  , message: "Found hole of type " <> (dropEnd 4 $ show marloweType)
+  , severity: markerSeverity "Hint"
+  , code: ""
+  , source: ""
+  }
+  where
+  dropEnd :: Int -> String -> String
+  dropEnd n = fromCodePointArray <<< Array.dropEnd n <<< toCodePointArray
+
+holeToMarkers :: MarloweHole -> Array IMarkerData
+holeToMarkers hole@(MarloweHole { name, marloweType, row, column }) =
+  let
+    m = getMarloweConstructors marloweType
+
+    constructors = take 1 $ Set.toUnfoldable $ Map.keys m
+  in
+    map (holeToMarker hole m) constructors
+
+markerToHole :: IMarkerData -> MarloweType -> MarloweHole
+markerToHole { startColumn, startLineNumber } marloweType = MarloweHole { name: "unknown", marloweType, row: startLineNumber, column: startColumn }
+
+warningToMarker :: Warning -> IMarkerData
+warningToMarker warning =
+  let
+    { startColumn, startLineNumber, endColumn, endLineNumber } = getWarningRange warning
+  in
+    { startColumn
+    , startLineNumber
+    , endColumn
+    , endLineNumber
+    , message: show warning
+    , severity: markerSeverity "Warning"
+    , code: ""
+    , source: ""
+    }
+
+format :: String -> String
+format contractString = case parseContract contractString of
+  Left _ -> contractString
+  Right contract -> show $ pretty contract
+
+provideCodeActions :: Uri -> Array IMarkerData -> Array CodeAction
+provideCodeActions uri markers' =
+  (flip foldMap) markers' \(marker@{ message, startLineNumber, startColumn, endLineNumber, endColumn }) -> case regex "Found hole of type (\\w+)" noFlags of
```

Comment:
Are you sure this is worth flipping?

Reply:
When you have a big function like this that is the main part of `foldMap` I like to because if you put the function by itself, the fact that it is being folded later makes it look like a weird function as it will return `m (m a)`. Just my thought.